builtin/providers/aws/resource_aws_cloudfront_web_distribution.go

```diff
@@ -0,0 +1,587 @@
+	if err != nil {
+		return err
+	}
+
+	params := &cloudfront.UpdateDistributionInput{
+		DistributionConfig: c,
+		Id:                 aws.String(string(d.Id())),
+		IfMatch:            v.ETag,
+	}
+
+	_, err = cloudfrontconn.UpdateDistribution(params)
+
+	if err != nil {
+		return fmt.Errorf("Error updating CloudFront distribution: %s", err)
+	}
+
+	err = resourceAwsCloudFrontWebDistributionWaitUntilDeployed(d, meta)
+	if err != nil {
+		return err
+	}
+
+	return resourceAwsCloudFrontWebDistributionRead(d, meta)
+}
+
+func resourceAwsCloudFrontWebDistributionDelete(d *schema.ResourceData, meta interface{}) error {
+	cloudfrontconn := meta.(*AWSClient).cloudfrontconn
+
+	v, err := resourceAwsCloudFrontWebDistributionDistributionRetrieve(d, meta)
+	if err != nil {
+		return err
+	}
+
+	// Do nothing if resource no longer exists
+	if v == nil {
+		return nil
+	}
+
+	// CloudFront distributions must be disabled in order to be deleted
+	if d.Get("enabled") == true {
+		d.Set("enabled", false)
+
+		err := resourceAwsCloudFrontWebDistributionUpdate(d, meta)
+		if err != nil {
+			return err
+		}
+
+		// Retrieve the latest ETag again
+		v, err = resourceAwsCloudFrontWebDistributionDistributionRetrieve(d, meta)
+		if err != nil {
+			return err
+		}
+	}
+
+	params := &cloudfront.DeleteDistributionInput{
+		Id:      aws.String(string(d.Id())),
+		IfMatch: v.ETag,
+	}
+
+	_, err = cloudfrontconn.DeleteDistribution(params)
+
+	if err != nil {
+		return fmt.Errorf("Error deleting CloudFront distribution: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsCloudFrontWebDistributionDistributionConfig(
+	d *schema.ResourceData, meta interface{},
+	callerReference *string) (*cloudfront.DistributionConfig, error) {
+
+	originId := fmt.Sprintf("%s-origin", d.Get("origin_domain_name"))
+	aliases := resourceAwsCloudFrontWebDistributionAwsStringLists(
+		d.Get("aliases"))
+	geoRestrictions := resourceAwsCloudFrontWebDistributionAwsStringLists(
+		d.Get("geo_restrictions"))
+	allowedMethods := resourceAwsCloudFrontWebDistributionHandleMethods(
+		d.Get("allowed_methods"))
+	cachedMethods := resourceAwsCloudFrontWebDistributionHandleMethods(
+		d.Get("cached_methods"))
+	forwardedHeaders := resourceAwsCloudFrontWebDistributionAwsStringLists(
+		d.Get("forwarded_headers"))
+	cookies := resourceAwsCloudFrontWebDistributionCookies(
+		d.Get("forward_cookie"), d.Get("whitelisted_cookies"))
+	viewerCertificate := &cloudfront.ViewerCertificate{
+		MinimumProtocolVersion: aws.String(d.Get("minimum_ssl").(string)),
+		SSLSupportMethod:       aws.String(d.Get("ssl_support_method").(string)),
+	}
+	if d.Get("certificate_id") == "" {
+		viewerCertificate.CloudFrontDefaultCertificate = aws.Bool(true)
+	} else {
+		viewerCertificate.IAMCertificateId = aws.String(d.Get("certificate_id").(string))
+	}
+
+	// PUT DistributionConfig requires, unlike POST, EVERY possible option to be set.
```

Comment:
Oh, I see, that answers my question below about `resourceAwsCloudFrontWebDistributionHandleMethods`. :cry: